pkg/apis/apps/validation/validation_test.go

```diff
@@ -1044,6 +1025,8 @@ func TestValidateStatefulSetUpdate(t *testing.T) {
 
 	for testName, errorCase := range errorCases {
 		t.Run(testName, func(t *testing.T) {
+			errorCase.old.ObjectMeta.ResourceVersion = "1"
```

Comment:
Good catch, can you change the current test to actually verify the contents of the errors instead of just their presences, something similar to https://github.com/kubernetes/kubernetes/blob/30adcd0b6c1c28303d7281b7be2ac45bdd7989e4/pkg/apis/batch/validation/validation_test.go#L737-L743 ?

Reply:
@soltysh upon farther inspection, found out that most of the negative tests in `TestValidateStatefulSetUpdate` are misleading. for example, https://github.com/kubernetes/kubernetes/blob/8fb2b901adce763853103dbebd56acfc827a1e84/pkg/apis/apps/validation/validation_test.go#L984-L1003 doesn't fail because of invalid selectors/labels, it fails because you're only allowed to modify `template`, `replicas`, etc.. in a statefulset. same with any other negative test in `TestValidateStatefulSetUpdate`. in fact, it looks like `ValidateStatefulSetUpdate` doesn't care about invalid labels or most of the other stuff that happen during the validation of a statefulset, it mostly verifies that during the modification of a statefulset, you didn't modify stuff that you aren't allowed to modify (e.g. PVC template)
(mostly, because it does verify that replicas and some other fields are not negative, not sure what's the reason behind that).

So I've modified the tests in `TestValidateStatefulSetUpdate` to be more explicit in what they're testing (e.g. `update pod management policy should fail`)

I've applied your suggestion (verifying the contents of the errors) to both `TestValidateStatefulSetUpdate` and `TestValidateStatefulSet`

Reply:
I've also changed some of the tests in `TestValidateStatefulSet` to fail for one (or few) reasons instead of multiple unrelated reasons

for example, `negative_replicas` was failing both because `replicas` was negative but also because it was missing the pod template

Reply:
This look so much better, thanks a lot for making the test much better!

Reply:
Setting resource version to pass update validation is a good catch.

> it looks like `ValidateStatefulSetUpdate` doesn't care about invalid labels or most of the other stuff that happen during the validation of a statefulset

hmm... that's pretty misleading, in my opinion. The only place we call ValidateStatefulSetUpdate (outside of tests) is from statefulSetStrategy#ValidateUpdate, where we also call ValidateStatefulSet.

It would be clearer and less error-prone if we moved the call to ValidateStatefulSet inside ValidateStatefulSetUpdate, rather than requiring two validation calls to get correct behavior, rather than removing/relaxing validation test cases like "empty pod creation policy" above.

I like the change to include the expected errors in the testcase, rather than just requiring failures to return _some_ error... that would have caught these scenarios failing with an error other than what we thought we were exercising.

Is there a way to separate out the whitespace-only non-functional test changes (like making the testcases a slice of named tests instead of a map) to their own commits so we can easily see the diffs that are changing test behavior?

Reply:
@liggitt 

> I like the change to include the expected errors in the testcase, rather than just requiring failures to return some error... that would have caught these scenarios failing with an error other than what we thought we were exercising.
> 
> Is there a way to separate out the whitespace-only non-functional test changes (like making the testcases a slice of named tests instead of a map) to their own commits so we can easily see the diffs that are changing test behavior?

I can replace the last commit with 3

1. making the tests a slice instead of a map with the ability to provide expected errors (at this point, I won't provide any expected errors)
2. changing the actual tests (e.g. adding `update pvc template storage class`, removing `invalid pod`)
3. adding the expected errors to the error cases

would that make it better?

> hmm... that's pretty misleading, in my opinion. The only place we call ValidateStatefulSetUpdate (outside of tests) is from statefulSetStrategy#ValidateUpdate, where we also call ValidateStatefulSet.
> 
> It would be clearer and less error-prone if we moved the call to ValidateStatefulSet inside ValidateStatefulSetUpdate, rather than requiring two validation calls to get correct behavior, rather than removing/relaxing validation test cases like "empty pod creation policy" above.

I wouldn't mind, but does it mean that we'll need to duplicate some tests, e.g. testing invalid labels both in `ValidateStatefulSet` and `ValidateStatefulSetUpdate`?

Reply:
> 1. making the tests a slice instead of a map with the ability to provide expected errors (at this point, I won't provide any expected errors)

yes, no-op structural change

> 2. changing the actual tests (e.g. adding update pvc template storage class, removing invalid pod)
> 3. adding the expected errors to the error cases

adding the expected errors first would be clearer to understand what coverage we're actually getting today

> I wouldn't mind, but does it mean that we'll need to duplicate some tests, e.g. testing invalid labels both in ValidateStatefulSet and ValidateStatefulSetUpdate?

no, not really... keeping the update tests focused on disallowed transitions is sufficient